### ISSUE TITLE
[representation] iS2S: Structure 2 Sequence Code Embeddings

### ISSUE THREAD

**Issue by danaderp** [WM-SEMERU/ds4se#94]
**Description**
Code Embeddings are abstract representations of source code employed in multiple automation tasks in software engineering like clone detection, traceability, or code generation. This abstract representation is a mathematical entity known as Tensor. Code Tensors allows us to manipulate snippets of code in semantic vector spaces instead of complex data structures like call graphs. Initial attempts focused on identifying deep learning strategies to compress code in lower-dimensional vectors (code2vec). Unfortunately, these approaches do not consider autoencoder architectures to represent code. The purpose of this project is to combine a structural language model of code with autoencoder architectures to compress source code snippets into lower-dimensional tensors. The lower-dimensional tensor must be evaluated in terms of semantics (clone detection).

**Project Goals**
- [ ] Check and analyze literature review in structural models in deep learning
- [ ] Implement a vanilla version of an autoencoder where the encoder is a structural language and the decoder a sequence-based architecture. 
- [ ] Lower dimensional tensors must be evaluated in terms of semantics (clone detection problem)

Implement a module of interpretability to test edge cases of the autoencoder

- Project Requirements
- Required Knowledge: Python, Git, and Statistics
- Preferred Knowledge: Deep Learning, TensorFlow, and DVC

Recommended Readings
- [Code2Vec](https://urialon.cswp.cs.technion.ac.il/wp-content/uploads/sites/83/2018/12/code2vec-popl19.pdf)
- [Code2Seq](https://openreview.net/pdf?id=H1gKYo09tX)
- [Structural Language Models](https://arxiv.org/pdf/1910.00577.pdf)
- [Keras Autoencoders](https://blog.keras.io/building-autoencoders-in-keras.html)

**Comment by danaderp** [comment]
For Sam:

- [ ] Run Code2Vec and checkout the architecture
- [ ] Read the three papers and discuss the results
- [ ] Set-up ds4se environment

**Comment by m13253** [comment]
Updated work plan:

- [X] Check and analyze literature review in structural models indeep learning
- [X] Set up DS4SE environment
- [X] Test out the original code2vec
- [ ] Using code2vec as a library to generate embedding vectors from CodeSearchNet-Java
- [ ] Use RNNs to build an autoencoder to compress the dimension of embedding vectors
- [ ] Move our autoencoder to use code2vec encoder
- [ ] Later, to use transformers decoder
- [ ] Evaluate our model for clone detection
- [ ] Test our autoencoder for interpretability by actively trans-forming the data and evaluate the output
- [ ] If time permits, extend the work to traceability and codegeneration

**Comment by m13253** [comment]
Meeting note 2021-03-18:
* Use the Keras autoencoder as a template to train against CodeSearchNet-Java.
* Use google/sentencepiece for input tokenization.
* Add an embedding layer after it.
* Use multiclass cross entropy for loss function (binary doesn't work).
* Don't deploy code2vec yet. It doesn't fit.

**Comment by m13253** [comment]
**Meeting note 2021-03-25:**
* Ignore excessive length snippets
* Prepare for "design of the case studies"

**Comment by m13253** [comment]
**Update 2021-04-01**:
* Finally got the autoencoder training running.
* The network is shrunk in its dimension to deal with OOM situations.
* Not seeing significant loss reduction.
* Need some help in "design of the case studies", if possible, a guideline in text version.
![image](https://user-images.githubusercontent.com/1808930/113356268-82149680-9331-11eb-8be9-5d514bbfab85.png)

**Comment by m13253** [comment]
**Meeting note 2021-04-02:**

#### Tasks:
1. Complete sampling
2. How to incorporate code2vec

#### Sampling (2 types):
1. focus on encoder, obtain the middle vectors (<- for now)
2. create random noise, how do they generate code

#### Case of studies:
The experiments we are going to run.
- Checking the clones (clone library provided)
- Test the encoder & decoder of GRU
- The other case: encoder is code2vec, decoder is GRU
